File: Analysis/ACE/_shared/load_ace_phenotype.py
"""Shared ACE phenotype loader for Tsai / DeJager workflows.

The canonical CSV (TSAI_DEJAGER_all_patients_wACEscores.csv) pools both
cohorts and contains duplicate projids. Any workflow that does
pheno.set_index("projid").reindex(cohort_ids) must filter to a single cohort
first or pandas raises ValueError("cannot reindex on an axis with duplicate
labels"). Use load_for_projids() as the single entry point.
"""

from __future__ import annotations

from pathlib import Path
from typing import Iterable, Union

import numpy as np
import pandas as pd

PathLike = Union[str, Path]

REQUIRED_COLUMNS = [
    "projid",
    "tot_adverse_exp",
    "early_hh_ses",
    "msex",
    "age_death",
    "pmi",
    "niareagansc",
]

ACE_COMPONENTS = [
    "emotional_neglect",
    "family_pro_sep",
    "financial_need",
    "parental_intimidation",
    "parental_violence",
]


def _add_ace_aggregate(pheno: pd.DataFrame) -> pd.DataFrame:
    if "ace_aggregate" in pheno.columns:
        return pheno
    if not {"tot_adverse_exp", "early_hh_ses"}.issubset(pheno.columns):
        return pheno
    z_adv = (pheno["tot_adverse_exp"] - pheno["tot_adverse_exp"].mean()) / pheno["tot_adverse_exp"].std()
    z_ses = (pheno["early_hh_ses"] - pheno["early_hh_ses"].mean()) / pheno["early_hh_ses"].std()
    pheno = pheno.copy()
    pheno["ace_aggregate"] = z_adv - z_ses
    return pheno


def load_raw(pheno_csv: PathLike) -> pd.DataFrame:
    """Load the ACE phenotype CSV with schema validation and ace_aggregate derived."""
    pheno = pd.read_csv(pheno_csv)
    # Drop the R-written rowname column if present ("" header → "Unnamed: 0").
    # Carrying it makes drop_duplicates() ineffective because every row has a
    # unique rownumber, leaving spurious projid duplicates in the pooled CSV.
    if "Unnamed: 0" in pheno.columns:
        pheno = pheno.drop(columns=["Unnamed: 0"])
    missing = set(REQUIRED_COLUMNS) - set(pheno.columns)
    if missing:
        raise ValueError(f"Phenotype CSV missing columns: {sorted(missing)}")
    pheno["projid"] = pheno["projid"].astype(str)
    return _add_ace_aggregate(pheno)


def load_for_projids(
    pheno_csv: PathLike,
    projids: Iterable[str],
    require_all: bool = False,
) -> pd.DataFrame:
    """Load phenotypes restricted to *projids* (e.g. pseudobulk obs_names).

    Filters to rows whose projid is in *projids*, drops exact duplicate rows,
    and raises if any projid still appears more than once (non-exact
    duplicate — indicates real data conflict, not pool-artifact duplication).

    Parameters
    ----------
    pheno_csv
        Path to the ACE phenotype CSV.
    projids
        Iterable of projid strings to keep.
    require_all
        If True, raise when any projid in *projids* is missing from the CSV.

    Returns
    -------
    DataFrame indexed by projid (str), ordered to match *projids*.
    """
    wanted = pd.Index([str(p) for p in projids], name="projid")
    pheno = load_raw(pheno_csv)

    sub = pheno[pheno["projid"].isin(wanted)].copy()
    sub = sub.drop_duplicates()

    dup_mask = sub["projid"].duplicated(keep=False)
    if dup_mask.any():
        offenders = sorted(sub.loc[dup_mask, "projid"].unique().tolist())
        raise ValueError(
            "Phenotype CSV has conflicting rows for projid(s): "
            f"{offenders[:10]}{' ...' if len(offenders) > 10 else ''} "
            "(exact duplicates were dropped; these remain after dedup)."
        )

    sub = sub.set_index("projid")
    sub = sub.reindex(wanted)

    if require_all and sub.isna().all(axis=1).any():
        missing = wanted[sub.isna().all(axis=1)].tolist()
        raise ValueError(
            f"{len(missing)} projid(s) from cohort not found in phenotype CSV: "
            f"{missing[:10]}{' ...' if len(missing) > 10 else ''}"
        )
    return sub

File: Analysis/ACE/_shared/load_ace_phenotype.R
#' Shared ACE phenotype loader for Tsai / DeJager R workflows.
#'
#' The canonical CSV pools Tsai + DeJager and has duplicate projids.
#' Use load_ace_for_projids() before any projid-indexed join/merge.

ACE_REQUIRED_COLUMNS <- c(
  "projid", "tot_adverse_exp", "early_hh_ses",
  "msex", "age_death", "pmi", "niareagansc"
)

.add_ace_aggregate <- function(pheno) {
  if ("ace_aggregate" %in% colnames(pheno)) return(pheno)
  if (!all(c("tot_adverse_exp", "early_hh_ses") %in% colnames(pheno))) return(pheno)
  z_adv <- scale(pheno$tot_adverse_exp)[, 1]
  z_ses <- scale(pheno$early_hh_ses)[, 1]
  pheno$ace_aggregate <- z_adv - z_ses
  pheno
}

load_ace_raw <- function(pheno_csv) {
  pheno <- read.csv(pheno_csv, stringsAsFactors = FALSE, check.names = FALSE)
  # Drop the R-written rowname column ("" header becomes "X" under base R's
  # read.csv). Carrying it breaks dedup because every row has a unique index.
  name_col_candidates <- c("X", "")
  drop_cols <- intersect(name_col_candidates, colnames(pheno))
  if (length(drop_cols) > 0) pheno <- pheno[, setdiff(colnames(pheno), drop_cols), drop = FALSE]
  missing_cols <- setdiff(ACE_REQUIRED_COLUMNS, colnames(pheno))
  if (length(missing_cols) > 0) {
    stop(
      "Phenotype CSV missing columns: ",
      paste(missing_cols, collapse = ", ")
    )
  }
  pheno$projid <- as.character(pheno$projid)
  .add_ace_aggregate(pheno)
}

#' Load phenotypes restricted to the supplied projids (e.g. cohort obs_names).
#'
#' Filters to rows whose projid is in `projids`, drops exact duplicate rows,
#' and stops if any projid remains duplicated after dedup.
#'
#' @param pheno_csv path to ACE phenotype CSV
#' @param projids character vector of projids to keep
#' @param require_all logical; stop if any projid in `projids` is missing
#' @return data.frame with rownames == projids (ordered to match `projids`)
load_ace_for_projids <- function(pheno_csv, projids, require_all = FALSE) {
  projids <- as.character(projids)
  pheno <- load_ace_raw(pheno_csv)

  sub <- pheno[pheno$projid %in% projids, , drop = FALSE]
  sub <- unique(sub)

  dup_ids <- sub$projid[duplicated(sub$projid) | duplicated(sub$projid, fromLast = TRUE)]
  if (length(dup_ids) > 0) {
    offenders <- unique(dup_ids)
    stop(
      "Phenotype CSV has conflicting rows for projid(s): ",
      paste(utils::head(offenders, 10), collapse = ", "),
      if (length(offenders) > 10) " ..." else "",
      " (exact duplicates were dropped; these remain after dedup)."
    )
  }

  rownames(sub) <- sub$projid
  sub <- sub[projids, , drop = FALSE]
  rownames(sub) <- projids

  if (require_all) {
    missing_pid <- projids[rowSums(!is.na(sub)) == 0]
    if (length(missing_pid) > 0) {
      stop(
        length(missing_pid), " projid(s) from cohort not found in phenotype CSV: ",
        paste(utils::head(missing_pid, 10), collapse = ", "),
        if (length(missing_pid) > 10) " ..." else ""
      )
    }
  }
  sub
}
